### ISSUE TITLE
Future of CSS Modules

### ISSUE THREAD

**Issue by taion** [css-modules/css-modules#187]
@geelen It looks like you're the primary maintainer here, but I assume your attention is mostly directed toward `styled-components` now. Who is the current active maintainer of this spec? If there isn't one, what steps can we take toward continued maintenance of this spec?

**Comment by geelen** [comment]
Yeah, I've not got terribly much time to put towards _either_ CSS Modules or Styled Components at the moment, but having both certainly doesn't help. Thanks for kicking off this discussion, I'll invite @markdalgleish, @sokra & @joshwnj to chime in as well.

Personally, I'm not working on particularly big CSS projects at the moment — [Front End Center](https://frontend.center) is my main focus rather than consulting/contracting these days. This means I'm not being exposed to some of the problems that are facing users of CSS Modules. Things like contextual overrides (the discussion in #147 I think has lead to some good candidates but I haven't had time to push them forward) are a good example. There are lots of small improvements (things like composition shorthands) that I'd love to see implemented, too. And then there's always the job of improving the documentation across the board, and quite a few PRs that could probably be merged straight away...

So for me, hacking away in pure-JS land is quite fine for the projects I'm playing with, and I'm definitely interested in exploring what is possible in that space, but I wouldn't suggest using _any_ CSS-in-JS tool over CSS Modules for decent-sized projects in the near term. So that means that there are real issues facing real projects here that are going unsolved by my absence, which is bad.

So as for "the future of CSS Modules", I think it would be best if I stepped back and let someone (or a few people) with the time and resources to steer it step forward. Any suggestions as to who that might include? Would you like to be involved, @taion?

**Comment by taion** [comment]
cc @jquense

I think it should definitely be more than one person. A key thing to the appeal of CSS Modules for us is that it's stable and relatively easy to understand – that it's relatively unopinionated, so it's a good default for getting modularization without being as cutting-edge as the CSS-in-JS approaches.

For my part, at least, I'm not looking for major changes – just considered incremental evolution and maintenance, and I think that's better with multiple people involved.

I see that @tivac has been pretty active on this repo. I'm not really sure what the landscape for CSS Modules looks like more broadly in terms of users.

**Comment by tivac** [comment]
I ❤️ the CSS Modules spec. I have a hard time wrapping my head around the code though. I'm not particularly interested in taking on any ownership of it, but I am **very** interested in the continuing spec/format discussions and future direction.

Definitely watching this thread closely. :+1:

**Comment by markdalgleish** [comment]
I have a very strong sense of guilt over my lack of involvement in this project due to my extremely busy schedule of work and family responsibilities.

That said, I'd love to get some more people involved, get some more active maintenance happening and keep this project moving forwards. I'm definitely keen to help with all of this, but I feel like we're missing an element of official coordination.

I propose that our next steps might be to build up a small group of maintainers, maybe start a development focused Slack channel (no questions from users, just coordinating work) and weekly video chat to keep us all on the page. Is anyone here interested in being a part of this?

**Comment by markdalgleish** [comment]
To add to that, I've been really inspired by the work of @TheLarkInn in building a really strong developer community around Webpack, where it used to be mostly a one-man show, at least as far as I could tell. Any input he can provide here would be really helpful, since there's a big overlap between the Webpack and CSS Modules communities.

**Comment by dvkndn** [comment]
I can help with issues and documentations. At my company, we are currently use CSS Modules heavily for almost all production products, some of them are decent-size (like a trading apps eco-system). Therefore, I think my hand is dirty enough with CSS Modules to help other people.

I actually did help some issues in the repo :D And it feel very great.

In the far future (or next company, or next pet project), I think I will jump to `styled-component` train. However, in the next near future, I'll stick a lot with CSS Modules

**Comment by sullenor** [comment]
Would like to mension that I (and my team) finally started to use CSS Module on our primary products. For example one them is a big web application for out partners. Unfortunately, I was busy with mirtating to the new technology stack for the past months and had not much time to help in here.

So would like to say that I'm interested in the future development of CSS Modules and would like to help here :) I'm not good with documenting stuff, but I can help with the code part.

I agree with @markdalgleish that *an element of official coordination* is missing. And a focus on the small working group with regular planning could help here. So I would like to join it also :)

**Comment by dfreeman** [comment]
I maintain the `ember-css-modules` integration, which my team uses extensively in our own projects and has also seen decent pickup in the wider Ember community. I suspect I don't have the capacity right now to take on a maintainer role here, but I'm definitely interested in seeing a slightly more formal group of maintainers come together.

I'd love to see public notes from the weekly meetings and/or (quietly) be a part of the development Slack in order to keep a finger on the pulse of things and see how they're developing :)

**Comment by taion** [comment]
@markdalgleish 

No need to feel guilty! CSS Modules is really good.

I think @TheLarkInn has done great work on the webpack side as well.

One contrast I'd like to draw, though, is that I think the CSS Modules spec is very close to complete – like on the order of 90% complete; there are only small things at the margin like #147 that need work, and all of that is a bit separate from the core of CSS Modules as I see it.

This is different from, say, webpack, which needed fundamental changes in v2. I think CSS Modules just needs a bit of housekeeping and moving forward small things at the margins to fill in the blanks is all.

@sullenor BTW, take a look at @jquense's [css-literal-loader](https://github.com/4Catalyzer/css-literal-loader)! 😆  It's been a nice little DX improvement for us.

**Comment by jquense** [comment]
:wave: I'm also interested in helping out. I think what @taion said echo's my own thoughts and wishes. We really dig CSS modules and would love to help make sure it says up to date, and stable, while also evolving it where needed!

**Comment by sullenor** [comment]
@taion I think that it requires *courage* to switch to the inlined css or smthing css-in-js like :)
Not sure that my team is prepared for that step 😄

Anyways it looks really good and I like the idea of using es6 template literals, so you can add a `css` function to inject styles in runtime without code compilation. Might be useful for the development stage. Also it is easily supported in tests and in the server-side rendering stuff.

Damn, I like it :)

**Comment by taion** [comment]
@sullenor If you look at @jquense's loader (assuming you're using webpack), it gives you both – the styles are pulled out at build time and treated as if they were separate CSS files. The code just happens to be inlined into the JS file, is all. So using that `css` template literal with css-literal-loader is the same as making a tiny .css file.

**Comment by joshwnj** [comment]
> I propose that our next steps might be to build up a small group of maintainers, maybe start a development focused Slack channel (no questions from users, just coordinating work) and weekly video chat to keep us all on the same page. Is anyone here interested in being a part of this?

I think this would be really valuable. Happy to pitch in where I can.

**Comment by kumarharsh** [comment]
It's nice to see this discussion being held, as it also gives new or potential adopters a chance to gauge the current scenario and future of this project. I'm just shifting my team from Radium to css-modules. Coming from the CSS-in-JS land, I think there are still a lot of issues which need to be solved for CSS-in-JS to mature, and there is still a good amount of instability in the libraries themselves, where maintainers have pivoted to newer things or the projects have run into some really difficult issues (https://github.com/FormidableLabs/radium/issues/788) (https://github.com/Khan/aphrodite/issues/8), which, after investing a lot of time into it, deflates confidence in the long-term viability of such solutions. CSS-Modules, on the other hand, seems much more stable, and it being as close to normal 'css' as possible also is a great plus.

I'd like to contribute to this community, but as a newcomer, I'll be happy to observe and learn more about the process and the spec itself.

**Comment by TheLarkInn** [comment]
Thank you @markdalgleish it's humbling to see these words and the efforts you are putting together here. As I'm already 200% overbooked on time myself with webpack, I can't unfortunately help in regards to this project, but I'm always willing to help give some advice or help however possible with the limited time I have.

**Comment by joeybaker** [comment]
I did a little work on the browserify implementation of css-modules (though have since switched to webpack). I once wrote a [post on CSS-in-JS not quite being the answer](https://byjoeybaker.com/react-inline-styles), but how CSS-modules probably were, so at the very least, I'm a fan! I would be happy to help where I can.

**Comment by taion** [comment]
@TheLarkInn I don't think we're asking for your time (: Just praising the work you've done with webpack and looking at it as a model.

**Comment by dvkndn** [comment]
Hey guys so what should we start doing now? I can go through our issues and close solved ones. I can also continue the work of this PR https://github.com/css-modules/css-modules/pull/65 to make it to the README

**Comment by augbog** [comment]
:+1: would like to see the css-modules spec move forward. I would also be interested in joining the Slack channel @markdalgleish proposed. Would someone like to get that started along with a [Slack invite page](https://github.com/outsideris/slack-invite-automation)? I definitely think it would be good if the community was more involved in this project.

@dvkndn I think it would be awesome if you added that into the README. I'm currently going through that issue and I'm sure many other people are too.

**Comment by TheSisb** [comment]
Also would like to see this repo live on; it's starting to grow stale.  Just started using CSS Modules on a project at work and loving them, would be willing to contribute moving forward.

**Comment by taion** [comment]
I think the next step is to go back to @joshwnj's suggestion – who's interested in participating?

I'm up for it.

Would we be okay with Discord, incidentally? It's less work to set up since we wouldn't need a Heroku invite thing.

**Comment by dvkndn** [comment]
Yes I agree. I'm in too. Let's start by having a place for discussion

**Comment by kumarharsh** [comment]
discord sounds fine too. I'm in

**Comment by joeybaker** [comment]
I went ahead and created a Discord server: https://discord.gg/m2QBEZ5

**Comment by outdooricon** [comment]
> I wouldn't suggest using any CSS-in-JS tool over CSS Modules for decent-sized projects in the near term

@geelen From this comment, do you not even think that `styled-components` should be used? Would you mind expanding on this more?

**Comment by Undistraction** [comment]
Was there any consensus on this? Will CSS Modules be developed going forward or is it being kicked into the long grass? No criticism meant - I know how hard it can be to find the time to work on things - but would be good to know if it has a future.

**Comment by joshwnj** [comment]
@Undistraction I think it has a future (I'm using it in production, and I know many others are too), but l imagine that future as being mainly fixing bugs that come up, rather than adding a whole bunch of new features.

Are there certain things you have in mind when thinking about the future of the project?

**Comment by taion** [comment]
It'd be nice to clean up the issue tracker, at least.

**Comment by jquense** [comment]
i'd love to get some time to clean up some stuff. tbh tho most of my continued issues with CSS modules is tooling, and a few small things around `compose`. Its pretty solid already. I'd love to get some time to work on the the last 10% tho.

**Comment by Undistraction** [comment]
@joshwnj @taion @jquense Thanks for replying. I've been meaning to take a look for a long while and only just found the time in the last couple of days to sit down and play with it. Really like it and it feels very natural and a no-brainer so thanks to all who contributed. I do have to say as someone arriving at this and looking for a way in it feels like there really needs to be single solid point of reference outlining:

- Where the project is and where it's going
- A clear outline of the API
- An overview of the ecosystem
- Ways in for people who are 1) Building a Webapp 2) Building a Webapp that renders on the server as well 3) Building a multipage site.

If it's really pretty much feature complete, then that is fantastic, but I think you need to underline that it is mature and stable rather than incomplete and abandoned. A big part of this is the fact that if I google CSS Modules I get this: https://github.com/css-modules/css-modules which hasn't been touched in 10 months and is minimal to say the least (for example the section on composition). 

I'm definitely concerned that people will be put off from using this which would be real shame as it is a great project.

**Comment by joshwnj** [comment]
> I think you need to underline that it is mature and stable rather than incomplete and abandoned

good call.  One of the ways I imagine us getting to this point is building up a list of tutorials or examples that demonstrate the different features. Because CSS Modules is more of an "idea" than a library, with many different libraries that implement the idea in different ways, it would be helpful for the README to reflect that diversity.

**Comment by MatteoWebDesigner** [comment]
There is a lot of focus on CSS in JS right now but I think CSS Module is still a good option. I have some proposal and I would love to learn more how CSS module works under.

**Comment by eduardoinnorway** [comment]
Any news here? We really like the concept and It would be really sad if it just fades away.

**Comment by taion** [comment]
We'd be up for getting involved and pushing forward some of the stalled proposals. cc @jquense @geelen

**Comment by mxl** [comment]
@geelen @markdalgleish Hey! Is there final solution regarding the discussion above?
[css-modules-loader-core](https://github.com/css-modules/css-modules-loader-core) seems not maintained. If someone is still interested in this project - we should come up with some solution without delay. As @markdalgleish [suggested](https://github.com/css-modules/css-modules/issues/187#issuecomment-257785081) you can start a development focused Slack channel. I'm sure that there are enough developers who can help to maintain css-modules organization projects and I'm one of them. And that's all we need to continue maintaining this wonderful project. I use css-modules in my projects and it's great idea and implementation and it'll be very sad if the project will just die. WDYT?

**Comment by kumarharsh** [comment]
+1

**Comment by tivac** [comment]
@mxl 

> I went ahead and created a Discord server: https://discord.gg/m2QBEZ5

has a few folks in it, but not much. No real dev discussions to speak of.

I'm 100% still interested in the ideas/syntax of CSS Modules, I'm just using them & iterating on them in my own project since this one seems pretty unloved.

**Comment by feluxe** [comment]
> There is a lot of focus on CSS in JS right now but I think CSS Module is still a good option.

I think it is the cleanest and most beautiful solution out there and the only one that allows me to decouple *theming* from the component entirely (just create a CSS file for each theme, load the one you need via CSS-Modules and pass it to the component via props). In this regard *styled-components* makes a step back, because each theme needs to be hard-coded into the component.

**Comment by Undistraction** [comment]
So what is the future of CSS Modules?

**Comment by TrySound** [comment]
Medium stories is official roadmap. We are waiting landing css as first class citizen in webpack.

**Comment by Undistraction** [comment]
@TrySound Thanks for the confirmation. And thanks for your work on CSS Modules. It would be great to add a brief Roadmap to the README so the project doesn't look like it's died.

**Comment by cowboy** [comment]
So... about the future of CSS Modules. Any status updates? It's a little disconcerting to see no real updates or movement (let alone anything that looks like a real set of best practices around naming conventions, composes, etc) in over a year.

**Comment by taion** [comment]
i’m not sure that doing this is the place of this repo, though. we can always separately work to promulgate our own best practices. for what we’ve been doing so far, CSS modules have just worked and continue to work

**Comment by jquense** [comment]
yeah I don't know that there is actually much more to do _here_. we've been using css-modules for a few years now and honestly it feels pretty done. It effectively scopes css. Other features like variables, etc are best left to Sass, et al.  The only thing I think folks really want but is missing is using other css-module classes as selectors. But that really needs a champion with time to make it happen.

**Comment by huan086** [comment]
My little contribution: now there's Typescript definitions for the various postcss-modules plugins. css-modules can thus be used in webpack.config.ts written Typescript.

Combined with [@jpavon/typings-for-css-modules-loader](https://www.npmjs.com/package/@jpavon/typings-for-css-modules-loader) to make Typescript definitions for the css-modules JSON export, css-modules is totally Typescript friendly now.

**Comment by vviikk** [comment]
I wrote a little article about [how styled-components is the next step in the evolution](https://medium.com/@piggyslasher/the-state-of-css-css-in-js-how-styled-components-is-solving-the-problems-weve-had-for-decades-d8abbc8bc148) of css modules. I think there's some useful info in there about the state of CSS. I was a heavy css modules user before moving to styled-components BTW. @geelen is one of the co-authors of that anyway :)

**Comment by hinell** [comment]
It's May 2023, The project is dead.  CSS in JS also got a lot of problems, chckout:
* [Why We're Breaking Up with CSS-in-JS](https://dev.to/srmagura/why-were-breaking-up-wiht-css-in-js-4g9b?utm_campaign=nodeland&utm_medium=email&utm_source=Revue%20newsletter)
